File: common/src/main/kotlin/net/impleri/playerskills/integrations/rei/SkillsDisplayVisibility.kt
package net.impleri.playerskills.integrations.rei

import dev.architectury.event.EventResult
import me.shedaniel.rei.api.client.registry.display.DisplayCategory
import me.shedaniel.rei.api.client.registry.display.visibility.DisplayVisibilityPredicate
import me.shedaniel.rei.api.common.display.Display
import me.shedaniel.rei.api.common.entry.EntryIngredient
import me.shedaniel.rei.api.common.entry.EntryStack
import net.impleri.playerskills.client.api.ItemRestrictionClient
import net.impleri.playerskills.events.ClientSkillsUpdatedEvent
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack

class SkillsDisplayVisibility : DisplayVisibilityPredicate {
  private val producibility: MutableMap<Item, Boolean> = HashMap()
  private val consumability: MutableMap<Item, Boolean> = HashMap()

  init {
    ClientSkillsUpdatedEvent.EVENT.register {
      clearCache()
    }
  }

  private fun clearCache() {
    PlayerSkillsLogger.ITEMS.debug("Clearing REI Visibility caches")
    producibility.clear()
    consumability.clear()
  }

  override fun getPriority(): Double {
    return 100.0
  }

  override fun handleDisplay(category: DisplayCategory<*>?, display: Display): EventResult {
    if (matchAnyIngredientInList(display.outputEntries) { hasHiddenOutput(it) }) {
      return EventResult.interruptFalse()
    }

    return if (matchAnyIngredientInList(display.inputEntries) { hasHiddenInput(it) }) {
      EventResult.interruptFalse()
    } else {
      EventResult.pass()
    }
  }

  private fun matchAnyIngredientInList(entries: List<EntryIngredient>, predicate: (EntryStack<*>) -> Boolean): Boolean {
    return entries.any {
      it.any(predicate)
    }
  }

  /**
   * Checks every ingredient to see if any are uncraftable
   */
  private fun isntProducible(item: Item): Boolean {
    return !producibility.computeIfAbsent(item, ItemRestrictionClient::canCraft)
  }

  private fun isntConsumable(item: Item): Boolean {
    return !consumability.computeIfAbsent(item, ItemRestrictionClient::canCraftWith)
  }

  private fun isFilteredAs(item: Item, predicate: (Item) -> Boolean): Boolean {
    return predicate(item)
  }

  private fun isFilteredAs(stack: ItemStack, predicate: (Item) -> Boolean): Boolean {
    return !stack.isEmpty && isFilteredAs(stack.item, predicate)
  }

  private fun hasHiddenOutput(entry: EntryStack<*>): Boolean {
    if (entry.isEmpty) {
      return false
    }

    return when (val value = entry.value) {
      is Item -> isFilteredAs(value) { isntProducible(it) }
      is ItemStack -> isFilteredAs(value) { isntProducible(it) }
      else -> false
    }
  }

  /**
   * Checks every ingredient to see if any are supposed to be hidden
   */
  private fun hasHiddenInput(entry: EntryStack<*>): Boolean {
    if (entry.isEmpty) {
      return false
    }
    return when (val value = entry.value) {
      is Item -> isFilteredAs(value) { isntConsumable(it) }
      is ItemStack -> isFilteredAs(value) { isntConsumable(it) }
      else -> false
    }
  }
}

File: common/src/main/scala/net/impleri/playerskills/integrations/rei/SkillsDisplayVisibility.scala
package net.impleri.playerskills.integrations.rei

import me.shedaniel.rei.api.client.registry.display.visibility.DisplayVisibilityPredicate
import me.shedaniel.rei.api.client.registry.display.DisplayCategory
import me.shedaniel.rei.api.client.registry.display.DisplayRegistry
import me.shedaniel.rei.api.common.display.Display
import me.shedaniel.rei.plugin.common.displays.brewing.{BrewingRecipe => RawBrewing}
import net.impleri.playerskills.api.restrictions.RestrictionType
import net.impleri.playerskills.client.restrictions.RecipeRestrictionOpsClient
import net.impleri.playerskills.integrations.rei.facades.BrewingRecipe
import net.impleri.playerskills.restrictions.RestrictionRegistry
import net.impleri.playerskills.restrictions.recipe.RecipeRestriction
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.impleri.slab.events.EventHandler
import net.impleri.slab.item.crafting.IsRecipe
import net.impleri.slab.item.crafting.Recipe
import net.impleri.slab.logging.Logger
import net.minecraft.world.item.crafting.{Recipe => RawRecipe}

import scala.collection.SeqView
import scala.collection.View
import scala.util.chaining.scalaUtilChainingOps

case class SkillsDisplayVisibility(
  displayRegistry: DisplayRegistry,
  restrictionRegistry: RestrictionRegistry = RestrictionRegistry(),
  recipeOpsClient: RecipeRestrictionOpsClient = RecipeRestrictionOpsClient(),
  logger: Logger = PlayerSkillsLogger.ITEMS,
) extends DisplayVisibilityPredicate
    with EventHandler {
  override def getPriority: Double = 100

  override def handleDisplay(
    category: DisplayCategory[_],
    display: Display,
  ) =
    failOn {
      castDisplayToRecipe(display)
        .fold(false)(hasMatchingRestriction)
        .pipe(Option(_))
    }

  private def getRestrictedRecipes: View[Recipe[_]] =
    restrictionRegistry.entries.view
      .filter(_.isType(RestrictionType.Recipe))
      .asInstanceOf[SeqView[RecipeRestriction]]
      .map(_.target)
      .filter(recipeOpsClient.isProducible(_, None))

  private def hasMatchingRestriction(recipe: IsRecipe): Boolean =
    recipe match {
      case r: Recipe[_] => getRestrictedRecipes.exists(_ == r)
      // TODO: Handle brewing
      case _ => false
    }

  private def castDisplayToRecipe(value: Display): Option[IsRecipe] =
    Option(displayRegistry.getDisplayOrigin(value)) flatMap {
      case r: RawRecipe[_] => Option(Recipe(r))
      case b: RawBrewing   => Option(BrewingRecipe(b))
      case _               => None
    }
}
